### PR TITLE
[BACKLOG-14324] VizAPI actions

### PR DESCRIPTION
**Pull request by dcleao**
@pentaho/millenniumfalcon, please review.

Depends on https://github.com/pentaho/pentaho-platform-plugin-common-ui/pull/962.

### CODE REVIEW

**Comment by dcleao** [comment]
@pentaho/millenniumfalcon Just a notice that I added a new commit with the changes to the event classes.

**Comment by dcleao** [comment]
@pentaho/millenniumfalcon Just a notice that I added a new commit with the VizAPI action classes!

**Comment by dcleao** [comment]
@pentaho/millenniumfalcon These commits finish up the raw coding work. Now need a week of testing and still some docs.

File: assemblies/platform-plugin/src/main/resources/resources/web/common-ui-require-js-cfg.js [hunk 1]
```diff
@@ -38,6 +38,7 @@
   // TODO: This fallback logic is temporary, and can be removed when the remaining
   //    parts of the system rename the "service" plugin id to "pentaho/service".
   var requireTypes = requireCfg.config["pentaho/service"] || (requireCfg.config["pentaho/service"] = {});
+  var requireTypeInfo = requireCfg.config["pentaho/typeInfo"] || (requireCfg.config["pentaho/typeInfo"] = {});
```
Comment (nantunes) [inline assemblies/platform-plugin/src/main/resources/resources/web/common-ui-require-js-cfg.js]: Why is this needed in this story?

File: impl/client/src/main/javascript/web/pentaho/type/action/base.js [hunk 2]
```diff
@@ -51,10 +51,43 @@ define([
...
+         * When the action is [asynchronous]{@link pentaho.type.action.Base.Type#isSync},
+         * this method _may_ return a promise. If the promise is rejected,
+         * the action is rejected with the rejection reason.
+         * However, if the promise is fulfilled, its value is always *ignored*.
+         *
+         * @param {pentaho.type.action.Base} action - The action.
+         *
+         * @return {?Promise} - A promise for the completion of default action of an asynchronous action, or `null`.
+         */
+        defaultAction: function(action) {
```
Comment (nantunes) [inline impl/client/src/main/javascript/web/pentaho/type/action/base.js]: The default action is so well separated I wonder why it exists.

File: impl/client/src/main/javascript/web/pentaho/visual/base/view.js [hunk 3]
```diff
@@ -729,6 +735,95 @@ define([
...
+      },
+      // endregion
+
+      // region Actions
+      __emitActionPhase: function(action, phase, isFinal) {
+        var eventType = action.type.id;
+
+        // TODO: emitGenericAsync when action is async.
+
+        this._emitGeneric(action, eventType, phase, isFinal ? null : _emitActionKeyArgs);
```
Comment (nantunes) [inline impl/client/src/main/javascript/web/pentaho/visual/base/view.js]: Where is `_emitGeneric` defined?

Comment (nantunes) [inline impl/client/src/main/javascript/web/pentaho/visual/base/view.js]: We might need to emit an event payload parallel to the action (like the original DOM event that triggered the action). Where do you see it fitting better?

File: impl/client/src/main/javascript/web/pentaho/type/action/base.js (hunk 2 above)
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/type/action/base.js]: Actually, I think this should be moved to beside the instance side protected methods.

File: impl/client/src/main/javascript/web/pentaho/visual/base/view.js (hunk 3 above)
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/visual/base/view.js]: It's `EventSource#_emitGeneric`.

Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/visual/base/view.js]: In principle, the action replaces the old event, for all uses. For that one, not sure. If, clearly, a DOM event triggered the action, then an action property would make sense to hold it. This would support async actions as well (once the DOM event is always accessible at window.event anyway).

File: assemblies/platform-plugin/src/main/resources/resources/web/common-ui-require-js-cfg.js (hunk 1 above)
Comment (dcleao) [inline assemblies/platform-plugin/src/main/resources/resources/web/common-ui-require-js-cfg.js]: This two lines, specifically:
```js
requireTypeInfo["pentaho/visual/action/select"] = {alias: "select"};
requireTypeInfo["pentaho/visual/action/execute"] = {alias: "execute"};
```
 allow attaching to action events using short ids. Check how `EventSource` resolves given eventType names by passing them through the `pentaho/typeInfo` service.

The other `pentaho/type` declarations in this file, I just added them for completeness.

File: impl/client/src/main/javascript/web/pentaho/type/action/base.js (hunk 2 above)
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/type/action/base.js]: And, it exists for implementing the actions' default behavior. Check SelectAction's implementation.

File: impl/client/src/main/javascript/web/pentaho/visual/base/view.js (hunk 3 above)
Comment (nantunes) [inline impl/client/src/main/javascript/web/pentaho/visual/base/view.js]: Sorry, the diff was hidden when I did the search.
